File: packages/sdk/src/domains/social.ts
/**
 * The social domain: friendships, blocks, and finding people.
 *
 * # The graph is server-owned, the client only mirrors it
 *
 * A friendship is a server-side edge between two accounts, and every mutation here — request,
 * respond, block — is a *request* that the server apply a rule it owns: who may message whom, whether
 * the recipient's privacy settings admit the request at all, whether either side is on the other's
 * block list. The client never derives relationship state locally, because a client-side guess would
 * drift from the server's the moment either party acted from another device. Instead the graph is
 * re-read ({@link listRelationships}) whenever {@link onFriendEvent} reports that something changed.
 *
 * # Discovery is prefix search, not a directory
 *
 * {@link search} matches a username prefix and {@link suggestions} returns accounts the graph thinks
 * are relevant (mutual friends). Both answer with {@link SuggestedUser} — public profile fields only.
 * Nothing here can enumerate accounts wholesale: a short or empty query is rejected or clamped
 * server-side, which is the enumeration defence a public directory would otherwise need.
 *
 * # Blocks are one-sided and silent
 *
 * {@link blockUser} sets a Block edge. Blocking is not a friendship teardown dialogue: the wire
 * carries no notification to the blocked party, and the block list is only ever visible to the
 * blocker through their own relationship list.
 *
 * # Mutes are quieter still
 *
 * {@link muteUser} sets a Mute edge — a personal, unnotified choice like a block, but lighter: it
 * neither tears down a friendship nor bars a message, it only marks an account the caller would
 * rather not hear from so the UI can hide that account's room chatter. The muted set is read back
 * as a filtered view of the relationship list ({@link mutedAccounts}), the same server-owned graph
 * a block lives in.
 */

import type { Id } from '@migo/wire';
import {
  OP,
  RelationshipKind,
  encodeFriendTarget,
  encodeFriendRespond,
  encodeRelationshipListReq,
  decodeRelationshipList,
  encodeSuggestReq,
  encodeSearchReq,
  encodeMuteSet,
  decodeSearchResponse,
  decodeFriendEvent,
  decodeAcknowledged,
} from '@migo/protocol';
import type {
  FriendEvent,
  FriendRespond,
  FriendTarget,
  MuteSet,
  RelationshipEntry,
  RelationshipListReq,
  SearchReq,
  SearchResponse,
  SuggestReq,
  SuggestedUser,
} from '@migo/protocol';

import { ListenerSet } from './listeners.js';
import type { Listener } from './listeners.js';
import type { EventErrorHandler, Rpc } from './rpc.js';

/**
 * The relationship page {@link listRelationships} asks for when the caller does not bound it.
 *
 * The wire takes a required limit, and the server treats it as a ceiling on the combined listing so a
 * large graph cannot flood a frame. Two hundred covers a personal social graph with room to spare;
 * a caller with a genuinely larger one pages explicitly.
 */
const DEFAULT_RELATIONSHIP_LIMIT = 200;

/**
 * Manage friendships and blocks, discover people, and observe friendship changes.
 *
 * One instance per client. {@link start} begins delivering {@link FriendEvent}s (which arrive on the
 * caller's own user topic, already subscribed by the client); the request methods work without it.
 */
export class SocialDomain {
  readonly #rpc: Rpc;
  readonly #friendListeners: ListenerSet<FriendEvent>;
  #unsubscribe: (() => void) | null = null;

  constructor(rpc: Rpc, onEventError?: EventErrorHandler) {
    this.#rpc = rpc;
    this.#friendListeners = new ListenerSet(OP.FRIEND_EVENT, onEventError);
  }

  /** Begins delivering friend events to registered handlers. Idempotent. */
  start(): void {
    if (this.#unsubscribe !== null) {
      return;
    }
    this.#unsubscribe = this.#rpc.on(OP.FRIEND_EVENT, decodeFriendEvent, (event) =>
      this.#friendListeners.deliver(event),
    );
  }

  /** Stops delivering friend events. Registered handlers are kept for a later {@link start}. */
  stop(): void {
    this.#unsubscribe?.();
    this.#unsubscribe = null;
  }

  /**
   * Registers a handler for friendship changes. Returns an unsubscribe function.
   *
   * An event names the other account and a `state` string (`"request"`, `"accepted"`); it is a hint
   * that the graph moved, not a source of truth — re-read {@link listRelationships} to draw the
   * right buttons, since the event carries no direction (incoming vs outgoing) and no removal state.
   */
  onFriendEvent(handler: Listener<FriendEvent>): () => void {
    return this.#friendListeners.add(handler);
  }

  /**
   * Sends a friend request.
   *
   * Resolves when the server has accepted the request *for delivery* — the recipient still has to
   * answer it, and a request the recipient's privacy settings forbid is rejected here with an error.
   * A repeated request while one is pending is idempotent server-side.
   */
  async friendRequest(userId: Id): Promise<void> {
    const request: FriendTarget = { userId };
    await this.#rpc.call(OP.FRIEND_REQUEST, encodeFriendTarget, decodeAcknowledged, request);
  }

  /**
   * Answers a pending friend request.
   *
   * `accept: false` declines (or withdraws an already-declined request); the edge simply disappears
   * from the graph either way. Only the *recipient* of a request may respond to it.
   */
  async friendRespond(userId: Id, accept: boolean): Promise<void> {
    const request: FriendRespond = { userId, accept };
    await this.#rpc.call(OP.FRIEND_RESPOND, encodeFriendRespond, decodeAcknowledged, request);
  }

  /**
   * Blocks an account.
   *
   * One-sided and unnotified: the blocked account is not told, and the block shows only in the
   * blocker's own {@link listRelationships}. Blocking also tears down any friendship between the two
   * server-side, so a caller that holds the relationship list should refresh it after this resolves.
   */
  async blockUser(userId: Id): Promise<void> {
    const request: FriendTarget = { userId };
    await this.#rpc.call(OP.BLOCK_SET, encodeFriendTarget, decodeAcknowledged, request);
  }

  /**
   * Mutes or unmutes an account for the caller.
   *
   * A personal, one-sided choice: `on: true` sets a Mute edge, `on: false` clears it, and neither
   * touches friendship or delivery — a muted account's messages still arrive, the client simply
   * hides that account's room chatter for the muter. Like a block it is unnotified and visible only
   * to the caller. A caller that holds the muted set should refresh it after this resolves.
   */
  async muteUser(userId: Id, on: boolean): Promise<void> {
    const request: MuteSet = { userId, on };
    await this.#rpc.call(OP.MUTE_SET, encodeMuteSet, decodeAcknowledged, request);
  }

  /**
   * Reads the accounts the caller has muted, as a plain list of ids.
   *
   * A thin projection over {@link listAllRelationships}: it reads the whole graph and keeps only the
   * Mute edges, since the wire carries mutes mixed in with friends, blocks, and the rest rather than
   * as a list of their own. The result is the set a client loads once at session start and consults
   * to hide muted voices.
   */
  async mutedAccounts(): Promise<Id[]> {
    const KIND_MUTE: number = RelationshipKind.Mute;
    const entries = await this.listAllRelationships();
    return entries.filter((entry) => entry.kind === KIND_MUTE).map((entry) => entry.userId);
  }

  /**
   * Reads the caller's relationship graph: friends, pending requests in both directions, follows,
   * and blocks, each as a {@link RelationshipEntry} whose `kind` is a {@link RelationshipKind} value.
   *
   * The list is the caller's own view (the block list of another account is never served), so a
   * caller re-reads it rather than maintaining a local mirror. `limit` bounds the combined result;
   * the default covers a personal graph.
   */
  async listRelationships(
    limit: number = DEFAULT_RELATIONSHIP_LIMIT,
  ): Promise<RelationshipEntry[]> {
    const request: RelationshipListReq = { limit };
    const response = await this.#rpc.call(
      OP.RELATIONSHIP_LIST,
      encodeRelationshipListReq,
      decodeRelationshipList,
      request,
    );
    return response.entries;
  }

  /**
   * Reads the caller's whole relationship graph in one unfiltered list: friends, pending
   * requests in both directions, follows and followers, blocks, and favourites.
   *
   * The wire is the same {@link RELATIONSHIP_LIST} call {@link listRelationships} makes, but the
   * client bounds nothing: `limit` rides as zero, which the server reads as "apply your own page",
   * so every kind the graph holds comes back mixed together and the *caller* filters by `kind`.
   * This is the form for a caller that wants the blocks and favourites alongside the friends
   * without naming a page size, where {@link listRelationships} is the form for a bounded read.
   */
  async listAllRelationships(): Promise<RelationshipEntry[]> {
    const request: RelationshipListReq = { limit: 0 };
    const response = await this.#rpc.call(
      OP.RELATIONSHIP_LIST,
      encodeRelationshipListReq,
      decodeRelationshipList,
      request,
    );
    return response.entries;
  }

  /**
   * Friend suggestions: accounts the graph considers relevant, strongest first.
   *
   * Each result carries a mutual-friend count, which is the only signal the server is willing to
   * expose about *why* an account was suggested. `limit` is clamped server-side; omit it for the
   * server's default page.
   */
  async suggestions(limit?: number): Promise<SuggestedUser[]> {
    const request: SuggestReq = {};
    if (limit !== undefined) {
      request.limit = limit;
    }
    const response: SearchResponse = await this.#rpc.call(
      OP.SUGGESTIONS,
      encodeSuggestReq,
      decodeSearchResponse,
      request,
    );
    return response.results;
  }

  /**
   * Searches public profiles by username prefix.
   *
   * The match is a prefix match on the username (display names are not searched), and the query is
   * required: there is deliberately no "list everyone" form, since that would be a directory dump
   * rather than a lookup. Results carry only public profile fields.
   */
  async search(query: string, limit?: number): Promise<SuggestedUser[]> {
    const request: SearchReq = { query };
    if (limit !== undefined) {
      request.limit = limit;
    }
    const response = await this.#rpc.call(
      OP.SEARCH,
      encodeSearchReq,
      decodeSearchResponse,
      request,
    );
    return response.results;
  }
}

File: clients/android/core/src/main/kotlin/com/migo/core/domain/Social.kt
package com.migo.core.domain

import com.migo.core.protocol.Acknowledged
import com.migo.core.protocol.FriendEvent
import com.migo.core.protocol.FriendRespond
import com.migo.core.protocol.FriendTarget
import com.migo.core.protocol.MuteSet
import com.migo.core.protocol.Op
import com.migo.core.protocol.RelationshipEntry
import com.migo.core.protocol.RelationshipKind
import com.migo.core.protocol.RelationshipListReq
import com.migo.core.protocol.SearchReq
import com.migo.core.protocol.SearchResponse
import com.migo.core.protocol.SuggestReq
import com.migo.core.protocol.SuggestedUser
import com.migo.core.wire.Id

/**
 * The social graph: friends, requests, blocks, people search, and suggestions.
 *
 * A port of `packages/sdk/src/domains/social.ts`. The graph is server-owned -- every mutation here
 * asks the server and the caller re-reads [listRelationships] afterwards, because a local mirror
 * would drift the moment either party acted from another device. [onFriendEvent] is the signal that
 * the graph moved; it says who moved, not how, so the answer to it is always a re-read.
 *
 * # Why the client never derives relationship state
 *
 * A friend request accepted from a second device must be reflected by this one, and the only honest
 * source is the server's own list. The event stream exists to tell the client *when* to look, never
 * to spare it the looking: [FriendEvent] carries no direction (incoming vs outgoing) and no removal
 * state, and a client that guessed from it would draw an Accept button on a request that was already
 * answered.
 *
 * # Search is a prefix match on usernames
 *
 * [search] matches username prefixes and returns public profile fields only; display names are not
 * searched. There is deliberately no "list everyone" form -- that would be a directory dump rather
 * than a lookup.
 */
class SocialDomain(
    private val rpc: Rpc,
    onEventError: EventErrorHandler? = null,
) {
    private val friendListeners = ListenerSet<FriendEvent>(Op.FRIEND_EVENT, onEventError)

    @Volatile
    private var subscription: Subscription? = null

    /**
     * Begins delivering friend events to registered handlers. Idempotent.
     *
     * Register handlers with [onFriendEvent] first: nothing is delivered before this is called, so
     * the ordering is what stops a client from missing the first event after connecting.
     */
    fun start() {
        if (subscription != null) return
        subscription = rpc.on(Op.FRIEND_EVENT, { r -> FriendEvent.decode(r) }) { e, _ ->
            friendListeners.deliver(e)
        }
    }

    /** Stops delivery. Registered handlers are kept for a later [start]. */
    fun stop() {
        val live = subscription ?: return
        subscription = null
        live.cancel()
    }

    /**
     * Registers a handler for friendship changes.
     *
     * An event names the other account and a `state` string (`"request"`, `"accepted"`); it is a hint
     * that the graph moved, not a source of truth -- re-read [listRelationships] to draw the right
     * buttons, since the event carries no direction (incoming vs outgoing) and no removal state.
     */
    fun onFriendEvent(listener: Listener<FriendEvent>): Subscription = friendListeners.add(listener)

    /**
     * Sends a friend request.
     *
     * Resolves when the server has accepted the request *for delivery* -- the recipient still has to
     * answer it, and a request the recipient's privacy settings forbid is rejected here with an error.
     * A repeated request while one is pending is idempotent server-side.
     */
    suspend fun friendRequest(userId: Id) {
        val request = FriendTarget(userId)
        rpc.call(Op.FRIEND_REQUEST, { w -> request.encode(w) }, { r -> Acknowledged.decode(r) })
    }

    /**
     * Answers a pending friend request.
     *
     * `accept = false` declines (or withdraws an already-declined request); the edge simply disappears
     * from the graph either way. Only the *recipient* of a request may respond to it.
     */
    suspend fun friendRespond(userId: Id, accept: Boolean) {
        val request = FriendRespond(userId, accept)
        rpc.call(Op.FRIEND_RESPOND, { w -> request.encode(w) }, { r -> Acknowledged.decode(r) })
    }

    /**
     * Blocks an account.
     *
     * One-sided and unnotified: the blocked account is not told, and the block shows only in the
     * blocker's own [listRelationships]. Blocking also tears down any friendship between the two
     * server-side, so a caller that holds the relationship list should refresh it after this resolves.
     */
    suspend fun blockUser(userId: Id) {
        val request = FriendTarget(userId)
        rpc.call(Op.BLOCK_SET, { w -> request.encode(w) }, { r -> Acknowledged.decode(r) })
    }

    /**
     * Mutes or unmutes one account for the caller.
     *
     * A personal choice, not a room's and not the other account's business. Like [blockUser] it is
     * one-sided and unnotified, but softer: a muted account's messages still arrive and simply carry a
     * mark the UI can honour, where a block severs the edge outright. `on = false` lifts the mute. The
     * mute shows only in the caller's own graph, so refresh [listMuted] after this resolves.
     */
    suspend fun muteUser(userId: Id, on: Boolean) {
        val request = MuteSet(userId, on)
        rpc.call(Op.MUTE_SET, { w -> request.encode(w) }, { r -> Acknowledged.decode(r) })
    }

    /**
     * Reads the caller's relationship graph: friends, pending requests in both directions, follows,
     * and blocks, each as a [RelationshipEntry] whose `kind` is a
     * [com.migo.core.protocol.RelationshipKind] value.
     *
     * The list is the caller's own view (the block list of another account is never served), so a
     * caller re-reads it rather than maintaining a local mirror. `limit` bounds the combined result;
     * the default covers a personal graph.
     */
    suspend fun listRelationships(limit: Long = DEFAULT_RELATIONSHIP_LIMIT): List<RelationshipEntry> {
        val request = RelationshipListReq(limit)
        val response = rpc.call(
            Op.RELATIONSHIP_LIST,
            { w -> request.encode(w) },
            { r -> com.migo.core.protocol.RelationshipList.decode(r) },
        )
        return response.entries
    }

    /**
     * Reads the caller's whole relationship graph in one unfiltered list.
     *
     * The wire is the same [listRelationships] call, but the client bounds nothing: `limit` rides as
     * zero, which the server reads as "apply your own page", so every kind the graph holds comes back
     * mixed together and the *caller* filters by `kind` -- the form for a caller that wants the blocks
     * and favourites alongside the friends without naming a page size.
     */
    suspend fun listAllRelationships(): List<RelationshipEntry> = listRelationships(limit = 0L)

    /**
     * The accounts the caller has muted, drawn from the one relationship graph.
     *
     * There is no separate "list mutes" call: mutes ride the same graph as friends and blocks, tagged
     * with [RelationshipKind.Mute], so this reads the whole graph through [listAllRelationships] and
     * keeps only that kind. Each returned entry's `kind` is the mute discriminant and its `userId` is
     * the muted account -- the field a caller drawing a Muted list actually wants.
     */
    suspend fun listMuted(): List<RelationshipEntry> {
        val muted = RelationshipKind.Mute.wire.toLong()
        return listAllRelationships().filter { it.kind == muted }
    }

    /**
     * Friend suggestions: accounts the graph considers relevant, strongest first.
     *
     * Each result carries a mutual-friend count, which is the only signal the server is willing to
     * expose about *why* an account was suggested. `limit` is clamped server-side; null for the
     * server's default page.
     */
    suspend fun suggestions(limit: Long? = null): List<SuggestedUser> {
        val request = SuggestReq(limit = limit)
        val response = rpc.call(
            Op.SUGGESTIONS,
            { w -> request.encode(w) },
            { r -> SearchResponse.decode(r) },
        )
        return response.results
    }

    /**
     * Searches public profiles by username prefix.
     *
     * The match is a prefix match on the username (display names are not searched), and the query is
     * required: there is deliberately no "list everyone" form, since that would be a directory dump
     * rather than a lookup. Results carry only public profile fields.
     */
    suspend fun search(query: String, limit: Long? = null): List<SuggestedUser> {
        val request = SearchReq(query, limit)
        val response = rpc.call(
            Op.SEARCH,
            { w -> request.encode(w) },
            { r -> SearchResponse.decode(r) },
        )
        return response.results
    }

    private companion object {
        /** The personal-graph page the bounded read asks for by default. */
        const val DEFAULT_RELATIONSHIP_LIMIT: Long = 100L
    }
}
